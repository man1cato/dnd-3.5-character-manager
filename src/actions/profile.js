import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
