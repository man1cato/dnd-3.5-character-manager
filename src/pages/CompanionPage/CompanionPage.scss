@import '../../styles/base/settings';


.CompanionPage__grid {
   display: grid;
   grid-template-columns: 1fr 3fr;
   grid-row-gap: 1rem;
}