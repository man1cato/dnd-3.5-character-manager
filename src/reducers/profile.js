export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
