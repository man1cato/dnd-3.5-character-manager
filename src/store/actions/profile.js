import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
