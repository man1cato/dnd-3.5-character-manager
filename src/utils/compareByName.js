export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}