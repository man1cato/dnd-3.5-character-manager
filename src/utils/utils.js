import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])