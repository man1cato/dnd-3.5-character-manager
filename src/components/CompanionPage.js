import React from 'react';

import { connect } from 'react-redux';
import update from 'react-addons-update';

import Header from './Header';
import Footer from './Footer';
import Abilities from './Abilities';
import Feats from './Feats';
import SpecialAbilities from './SpecialAbilities';
import SkillSet from './SkillSet';
import PhysicalStats from './PhysicalStats';
import Weapons from './Weapons';
import { startEditProfile } from '../actions/profile';


export class CompanionPage extends React.Component {
    state = {
        hp: {
            base: this.props.hp.base,
            mod: this.props.hp.mod || "",
            damage: this.props.hp.damage || "",
            total: this.props.hp.total || this.props.hp.base,
        },
        initiative: {
            base: this.props.initiative.base, 
            mod: this.props.initiative.mod || "",
            total: this.props.initiative.total || this.props.initiative.base
        }, 
        abilities: this.props.abilities
    }

    onInputChange = (e) => {
        const name = e.target.name;
        const id = e.target.id;
        let value = Number(e.target.value);
        value = (value === 0 || isNaN(value)) ? "" : value;
        let tempMod;
        let total;

        this.setState((prevState) => {
            if (name === "hp") {
                if (id === "damage") {
                    total = this.props.hp.base + prevState.hp.mod - value;
                    return {
                        hp: update(prevState.hp, {
                            damage: { $set: value },
                            total: { $set: total }
                        })
                    }
                } 
                total = this.props.hp.base + value - prevState.hp.damage;
            } else if (name === "saves" || name === "attacks") {
                total = this.props[name][id].base + value;
                return {
                    [name]: update(prevState[name], {
                      [id]: {
                        mod: { $set: value },
                        total: { $set: total }
                      }
                    })
                  }
            } else if (name === "abilities") {
                tempMod = value === "" ? "" : Math.floor(value/2 - 5);
                return {
                    abilities: update(prevState.abilities, {
                        [id]: {
                            tempScore: {$set: value},
                            tempMod: {$set: tempMod}
                        }
                    })
                }
            } else {
                total = this.props[name].base + value;
            }
            return {
                [name]: update(prevState[name], {
                    mod: { $set: value },
                    total: { $set: total }          
                })
            }
        });
    }

    componentWillUnmount() {
        this.props.startEditProfile(this.props.id, {
            hp: this.state.hp,
            initiative: this.state.initiative,
            abilities: this.state.abilities
        });
    }

    render() {
        return (
            <div className="layout">
                <Header pageTitle="Companion" />
                <div className="container container--body">
                    <div className="grid grid--companion">
                        <h3>{this.props.name}</h3>
                        <div>{this.props.type}</div>
                        <h4 className="grid__col1">Special Abilities</h4>
                        <SpecialAbilities specialAbilityIds={this.props.specialAbilities} />                    
                        <h4 className="grid__col1">Feats</h4>
                        <Feats featIds={this.props.feats} />
                    </div>

                    <Abilities
                        abilities={this.state.abilities}
                        onInputChange={this.onInputChange}
                    />

                    <SkillSet 
                        skillSet={this.props.skillSet}                        
                    />
                    
                    <h3 className="row row--center">Combat</h3>
                    <PhysicalStats
                        hp={this.state.hp}
                        ac={this.props.ac}
                        initiative={this.state.initiative}
                        speed={this.props.speed}
                        onInputChange={this.onInputChange}
                    />

                    <div className="grid grid--combat">
                        <div className="row__title">Attack</div>
                        <div className="row row--left">            
                                {this.props.attack}
                        </div>
                    </div>

                </div>
                <Footer />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    id: state.profile.id,
    ...state.profile.companion
})

const mapDispatchToProps = (dispatch, props) => ({
    startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
});


export default connect(mapStateToProps, mapDispatchToProps)(CompanionPage)