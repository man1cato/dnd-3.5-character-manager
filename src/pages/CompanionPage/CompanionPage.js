import React from 'react'
import { connect } from 'react-redux'

import Abilities from '../../components/Abilities/Abilities'
import Feats from '../../components/Feats/Feats'
import SpecialAbilities from '../../components/SpecialAbilities/SpecialAbilities'
import SkillSet from '../../components/SkillSet/SkillSet'
import PhysicalStats from '../../components/PhysicalStats/PhysicalStats'
import { startEditProfile } from '../../store/actions/profile'
import { calcAbilityMod } from '../../utils/utils'


export const CompanionPage = ({ id, companion, startEditProfile }) => {
	const handleUpdate = (updates) => {
		startEditProfile(id, {
			companion: {
				...companion,
				...updates
			}
		})
	}

	return (
		<div className="container--body">
			<div className="section grid--companion">
				<h3>{companion.name}</h3>
				<div>{companion.type}</div>

				<h4 className="grid__col1">Special Abilities</h4>
				<SpecialAbilities 
					specialAbilityIds={companion.specialAbilities} 
				/>                    
				
				<h4 className="grid__col1">Feats</h4>
				<Feats 
					featIds={companion.feats} 
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Abilities</h3>
				<Abilities
					abilities={companion.abilities}
					handleUpdate={handleUpdate}
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Skills</h3>
				<SkillSet 
					skillSet={companion.skillSet}                        
				/>
			</div>

			<h3 className="section__title">Combat</h3>
			<PhysicalStats
				hp={companion.hp}
				ac={companion.ac}
				initBase={
					!!companion.abilities.dex.tempScore ?
					calcAbilityMod(companion.abilities.dex.tempScore) :
					calcAbilityMod(companion.abilities.dex.score)
				}
				initMod={companion.initMod}
				speed={companion.speed}
				handleUpdate={handleUpdate}							
			/>

			<div className="grid--combat">
				<div className="row__title">Attack</div>
				<div className="row--left">            
					{companion.attack}
				</div>
			</div>

		</div>
	)
	
}

const mapStateToProps = (state) => ({
	id: state.profile.id,
	companion: state.profile.companion
})

const mapDispatchToProps = (dispatch) => ({
	startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
})


export default connect(mapStateToProps, mapDispatchToProps)(CompanionPage)