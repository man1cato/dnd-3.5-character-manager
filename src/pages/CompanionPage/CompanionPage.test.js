import React from 'react'

import { CompanionPage } from './CompanionPage'
import profile from '../../tests/fixtures/profile'
import companion from '../../tests/fixtures/companion'
import { apiData, renderWithRedux } from '../../tests/utils'


const startEditProfile = jest.fn()
let props, state
beforeEach(async () => {
    const api = await apiData()
    props = {
        id: profile.id,
        companion,
        startEditProfile
    }
    state = {
        specialAbilities: api.specialAbilities,
        feats: api.feats,
        skills: api.skills
    }
})


test('should render CompanionPage with profile data', () => {
    const { container } = renderWithRedux(<CompanionPage {...props} />, state)
    expect(container.firstChild).toMatchSnapshot()
})