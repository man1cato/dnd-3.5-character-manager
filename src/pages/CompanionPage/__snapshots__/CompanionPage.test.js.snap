// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CompanionPage with profile data 1`] = `
<div
  className="container--body"
>
  <div
    className="section grid--companion"
  >
    <h3>
      Nyx
    </h3>
    <div>
      Raven
    </div>
    <h4
      className="grid__col1"
    >
      Special Abilities
    </h4>
    <Connect(SpecialAbilities)
      specialAbilityIds={
        Array [
          "recoVjIceMNsVyiqI",
          "recC7zmmkwPjM104d",
          "recD7wgtCKPM9k7X9",
          "recmKq1KJyCGILkHN",
        ]
      }
    />
    <h4
      className="grid__col1"
    >
      Feats
    </h4>
    <Connect(Feats)
      featIds={
        Array [
          "recvzMqdCBIzd7z9n",
          "recVdnKmMFz54fNLF",
        ]
      }
    />
  </div>
  <div
    className="section"
  >
    <h3
      className="section__title"
    >
      Abilities
    </h3>
    <Abilities
      abilities={
        Object {
          "cha": Object {
            "score": 6,
          },
          "con": Object {
            "score": 10,
          },
          "dex": Object {
            "score": 15,
          },
          "int": Object {
            "score": 8,
          },
          "str": Object {
            "score": 1,
          },
          "wis": Object {
            "score": 14,
          },
        }
      }
      handleUpdate={[Function]}
    />
  </div>
  <div
    className="section"
  >
    <h3
      className="section__title"
    >
      Skills
    </h3>
    <Connect(SkillSet)
      skillSet={
        Array [
          Object {
            "name": "Hide",
            "ranks": 10,
          },
          Object {
            "name": "Climb",
            "ranks": 3,
          },
          Object {
            "name": "Listen",
            "ranks": 5,
          },
          Object {
            "name": "Concentration",
            "ranks": 9,
          },
          Object {
            "name": "Spot",
            "ranks": 5,
          },
        ]
      }
    />
  </div>
  <h3
    className="section__title"
  >
    Combat
  </h3>
  <PhysicalStats
    ac={
      Object {
        "base": 14,
        "flat": 12,
        "touch": 14,
      }
    }
    handleUpdate={[Function]}
    hp={
      Object {
        "base": 16,
      }
    }
    initBase={2}
    speed={
      Object {
        "flight": 40,
        "ground": 10,
      }
    }
  />
  <div
    className="grid--combat"
  >
    <div
      className="row__title"
    >
      Attack
    </div>
    <div
      className="row--left"
    >
      Claws +4 melee (1d2â€“5)
    </div>
  </div>
</div>
`;
